src/circuit_builder.rs

```diff
@@ -68,61 +54,71 @@ where
         &mut self,
         pub_params: &PublicParameters,
         prover_key: &ProverKey,
-        inputs: CircuitInputs,
         transcript_initialisation: &'static [u8],
-    ) -> Result<Proof, Error>;
+    ) -> Result<Proof>;
 
     /// Verifies a proof using the provided `CircuitInputs` & `VerifierKey` instances.
     fn verify_proof(
-        &self,
+        &mut self,
         pub_params: &PublicParameters,
         verifier_key: &VerifierKey,
         transcript_initialisation: &'static [u8],
         proof: &Proof,
         pub_inputs: &[PublicInput],
-    ) -> Result<(), Error>;
+    ) -> Result<()>;
+}
+
+/// Represents an error in the PublicParameters creation and or modification.
+#[derive(Error, Debug)]
+pub enum CircuitErrors {
+    /// This error occurs when the circuit is not provided with all of the
+    /// required inputs.
+    #[error("missing inputs for the circuit")]
+    CircuitInputsNotFound,
+    /// This error occurs when we want to verify a Proof but the pi_constructor
+    /// attribute is unninitialized.
+    #[error("PI constructor attribute is unninitialized")]
+    UnninitializedPIGenerator,
```

Comment:
Nit: fix typo to `UninitializedPIGenerator`, and change `unninitialized` to `uninitialized` in the error description